Sources/_CryptoExtras/RSA/RSA_security.swift

```diff
@@ -0,0 +1,320 @@
+    }
+
+    init<Bytes: DataProtocol>(derRepresentation: Bytes) throws {
+        let keyAttributes: [CFString: Any] = [
+            kSecAttrKeyType: kSecAttrKeyTypeRSA,
+            kSecAttrKeyClass: kSecAttrKeyClassPublic,
+        ]
+        let data = Data(derRepresentation)
+        var error: Unmanaged<CFError>? = nil
+        let key = SecKeyCreateWithData(data as CFData, keyAttributes as CFDictionary, &error)
+
+        if let error = error {
+            throw error.takeRetainedValue() as Error
+        }
+
+        guard let unwrappedKey = key else {
+            throw CryptoKitError.incorrectParameterSize
+        }
+
+        self.backing = unwrappedKey
+    }
+
+    var derRepresentation: Data {
+        var error: Unmanaged<CFError>? = nil
+        let representation = SecKeyCopyExternalRepresentation(self.backing, &error)
+        precondition(error == nil)
+        return representation! as Data
+    }
+
+    var pemRepresentation: String {
+        return ASN1.PEMDocument(type: "PUBLIC KEY", derBytes: self.derRepresentation).pemString
+    }
+
+    var keySizeInBits: Int {
+        let attributes = SecKeyCopyAttributes(self.backing)! as NSDictionary
+        return (attributes[kSecAttrKeySizeInBits]! as! NSNumber).intValue
+    }
+
+    fileprivate init(_ backing: SecKey) {
+        self.backing = backing
+    }
+}
+
+
+internal struct SecurityRSAPrivateKey {
+    private var backing: SecKey
+
+    static let PKCS1KeyType = "RSA PRIVATE KEY"
+
+    static let PKCS8KeyType = "PRIVATE KEY"
+
+    init(pemRepresentation: String) throws {
+        let document = try ASN1.PEMDocument(pemString: pemRepresentation)
+
+        switch document.type {
+        case SecurityRSAPrivateKey.PKCS1KeyType:
+            // This is what is expected by Security.framework
+            self = try .init(derRepresentation: document.derBytes)
+        case SecurityRSAPrivateKey.PKCS8KeyType:
+            guard let pkcs8Bytes = document.derBytes.pkcs8RSAKeyBytes else {
+                throw _CryptoRSAError.invalidPEMDocument
+            }
+            self = try .init(derRepresentation: pkcs8Bytes)
+        default:
+            throw _CryptoRSAError.invalidPEMDocument
+        }
+
+    }
+
+    init<Bytes: DataProtocol>(derRepresentation: Bytes) throws {
+        let keyAttributes: [CFString: Any] = [
+            kSecAttrKeyType: kSecAttrKeyTypeRSA,
+            kSecAttrKeyClass: kSecAttrKeyClassPrivate,
+        ]
+        let data = Data(derRepresentation)
+        var error: Unmanaged<CFError>? = nil
+
+        // We can't know in DER if this is PKCS8 or PKCS1 without just trying to decode it.
+        let keyData: Data
+        if let pkcs8Data = data.pkcs8RSAKeyBytes {
+            keyData = pkcs8Data
+        } else {
+            keyData = data
+        }
+
+        let key = SecKeyCreateWithData(keyData as CFData, keyAttributes as CFDictionary, &error)
+
+        if let error = error {
```

Comment:
this needs double checking. Typically in the ObjC/CF APIs you are only allowed to look at the error value if you have previously determined that there was an error. Given that you `nil` initialise this, this is likely totally fine, although I believe that you're not supposed to even look at the `NSError **/CFError **` output if you don't know (from the return value) that there was an error.

Reply:
Yes, this seems like good advice. I'll refactor this to check the return code first.